File: mobile/lib/models/message.dart
import 'event.dart';
import 'queue.dart';

class CloseCode {
  static const int unauthorized = 4001;
  static const int schemaMismatch = 4002;
  static const int slowClient = 4003;
}

enum PipelineMode {
  rawLines('raw_lines'),
  rawEvents('raw_events'),
  agentdesk('agentdesk');

  const PipelineMode(this.wireName);
  final String wireName;

  static PipelineMode fromJson(String value) {
    switch (value) {
      case 'raw_lines':
        return PipelineMode.rawLines;
      case 'raw_events':
        return PipelineMode.rawEvents;
      case 'agentdesk':
        return PipelineMode.agentdesk;
      default:
        throw ArgumentError('Unknown pipeline mode: $value');
    }
  }

  String toJson() => wireName;
}

enum TransportMode {
  tls('tls'),
  insecureDev('insecure_dev');

  const TransportMode(this.wireName);
  final String wireName;

  static TransportMode fromJson(String value) {
    switch (value) {
      case 'tls':
        return TransportMode.tls;
      case 'insecure_dev':
        return TransportMode.insecureDev;
      default:
        throw ArgumentError('Unknown transport mode: $value');
    }
  }

  String toJson() => wireName;
}

enum Decision {
  approve('approve'),
  deny('deny');

  const Decision(this.wireName);
  final String wireName;

  static Decision fromJson(String value) {
    switch (value) {
      case 'approve':
        return Decision.approve;
      case 'deny':
        return Decision.deny;
      default:
        throw ArgumentError('Unknown decision: $value');
    }
  }

  String toJson() => wireName;
}

enum CommandError {
  noSuchEvent('no_such_event'),
  notARequest('not_a_request'),
  alreadyResolved('already_resolved'),
  invalid('invalid');

  const CommandError(this.wireName);
  final String wireName;

  static CommandError fromJson(String value) {
    switch (value) {
      case 'no_such_event':
        return CommandError.noSuchEvent;
      case 'not_a_request':
        return CommandError.notARequest;
      case 'already_resolved':
        return CommandError.alreadyResolved;
      case 'invalid':
        return CommandError.invalid;
      default:
        throw ArgumentError('Unknown command error: $value');
    }
  }

  String toJson() => wireName;
}

const int logOffsetTail = -1;

class LogLine {
  final int offset;
  final DateTime ts;
  final String text;

  const LogLine({required this.offset, required this.ts, required this.text});

  factory LogLine.fromJson(Map<String, dynamic> json) {
    return LogLine(
      offset: (json['offset'] as num).toInt(),
      ts: DateTime.parse(json['ts'] as String).toUtc(),
      text: json['text'] as String,
    );
  }

  Map<String, dynamic> toJson() => {
    'offset': offset,
    'ts': ts.toUtc().toIso8601String(),
    'text': text,
  };

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is LogLine &&
          runtimeType == other.runtimeType &&
          offset == other.offset &&
          ts.isAtSameMomentAs(other.ts) &&
          text == other.text;

  @override
  int get hashCode => offset.hashCode ^ text.hashCode;
}

// Handshake
class Hello {
  final String token;
  final String deviceId;
  final String clientVersion;
  final int schemaVersion;

  const Hello({
    required this.token,
    required this.deviceId,
    required this.clientVersion,
    required this.schemaVersion,
  });

  factory Hello.fromJson(Map<String, dynamic> json) {
    return Hello(
      token: json['token'] as String,
      deviceId: json['device_id'] as String,
      clientVersion: json['client_version'] as String,
      schemaVersion: (json['schema_version'] as num).toInt(),
    );
  }

  Map<String, dynamic> toJson() => {
    'token': token,
    'device_id': deviceId,
    'client_version': clientVersion,
    'schema_version': schemaVersion,
  };
}

class Welcome {
  final String daemonVersion;
  final int schemaVersion;
  final PipelineMode pipelineMode;
  final TransportMode transport;
  final DateTime serverTime;

  const Welcome({
    required this.daemonVersion,
    required this.schemaVersion,
    required this.pipelineMode,
    required this.transport,
    required this.serverTime,
  });

  factory Welcome.fromJson(Map<String, dynamic> json) {
    return Welcome(
      daemonVersion: json['daemon_version'] as String,
      schemaVersion: (json['schema_version'] as num).toInt(),
      pipelineMode: PipelineMode.fromJson(json['pipeline_mode'] as String),
      transport: TransportMode.fromJson(json['transport'] as String),
      serverTime: DateTime.parse(json['server_time'] as String).toUtc(),
    );
  }

  Map<String, dynamic> toJson() => {
    'daemon_version': daemonVersion,
    'schema_version': schemaVersion,
    'pipeline_mode': pipelineMode.toJson(),
    'transport': transport.toJson(),
    'server_time': serverTime.toUtc().toIso8601String(),
  };
}

class Snapshot {
  final List<QueueEntry> entries;
  final List<Event> events;

  const Snapshot({required this.entries, required this.events});

  factory Snapshot.fromJson(Map<String, dynamic> json) {
    return Snapshot(
      entries:
          (json['entries'] as List<dynamic>?)
              ?.map(
                (e) => QueueEntry.fromJson((e as Map).cast<String, dynamic>()),
              )
              .toList() ??
          const [],
      events:
          (json['events'] as List<dynamic>?)
              ?.map((e) => Event.fromJson((e as Map).cast<String, dynamic>()))
              .toList() ??
          const [],
    );
  }

  Map<String, dynamic> toJson() => {
    'entries': entries.map((e) => e.toJson()).toList(),
    'events': events.map((e) => e.toJson()).toList(),
  };
}

class EventPush {
  final Event event;
  final QueueEntry entry;

  const EventPush({required this.event, required this.entry});

  factory EventPush.fromJson(Map<String, dynamic> json) {
    return EventPush(
      event: Event.fromJson((json['event'] as Map).cast<String, dynamic>()),
      entry: QueueEntry.fromJson(
        (json['entry'] as Map).cast<String, dynamic>(),
      ),
    );
  }

  Map<String, dynamic> toJson() => {
    'event': event.toJson(),
    'entry': entry.toJson(),
  };
}

class ScoreUpdate {
  final String eventId;
  final int score;
  final int escalationLevel;

  const ScoreUpdate({
    required this.eventId,
    required this.score,
    required this.escalationLevel,
  });

  factory ScoreUpdate.fromJson(Map<String, dynamic> json) {
    return ScoreUpdate(
      eventId: json['event_id'] as String,
      score: (json['score'] as num).toInt(),
      escalationLevel: (json['escalation_level'] as num).toInt(),
    );
  }

  Map<String, dynamic> toJson() => {
    'event_id': eventId,
    'score': score,
    'escalation_level': escalationLevel,
  };
}

class StateUpdate {
  final String eventId;
  final EntryState state;
  final Resolution? resolution;
  final bool superseded;

  const StateUpdate({
    required this.eventId,
    required this.state,
    this.resolution,
    this.superseded = false,
  });

  factory StateUpdate.fromJson(Map<String, dynamic> json) {
    return StateUpdate(
      eventId: json['event_id'] as String,
      state: EntryState.fromJson(json['state'] as String),
      resolution: json['resolution'] != null
          ? Resolution.fromJson(json['resolution'] as String)
          : null,
      superseded: json['superseded'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
    'event_id': eventId,
    'state': state.toJson(),
    if (resolution != null) 'resolution': resolution!.toJson(),
    'superseded': superseded,
  };
}

class RawLine {
  final String agentId;
  final int offset;
  final DateTime ts;
  final String text;

  const RawLine({
    required this.agentId,
    required this.offset,
    required this.ts,
    required this.text,
  });

  factory RawLine.fromJson(Map<String, dynamic> json) {
    return RawLine(
      agentId: json['agent_id'] as String,
      offset: (json['offset'] as num).toInt(),
      ts: DateTime.parse(json['ts'] as String).toUtc(),
      text: json['text'] as String,
    );
  }

  Map<String, dynamic> toJson() => {
    'agent_id': agentId,
    'offset': offset,
    'ts': ts.toUtc().toIso8601String(),
    'text': text,
  };
}

class EventRef {
  final String eventId;

  const EventRef({required this.eventId});

  factory EventRef.fromJson(Map<String, dynamic> json) {
    return EventRef(eventId: json['event_id'] as String);
  }

  Map<String, dynamic> toJson() => {'event_id': eventId};
}

class GetEventLogs {
  final String eventId;
  final int offset;
  final int limit;

  const GetEventLogs({
    required this.eventId,
    required this.offset,
    required this.limit,
  });

  factory GetEventLogs.fromJson(Map<String, dynamic> json) {
    return GetEventLogs(
      eventId: json['event_id'] as String,
      offset: (json['offset'] as num).toInt(),
      limit: (json['limit'] as num).toInt(),
    );
  }

  Map<String, dynamic> toJson() => {
    'event_id': eventId,
    'offset': offset,
    'limit': limit,
  };
}

class EventLogs {
  final String eventId;
  final int offset;
  final int total;
  final bool evicted;
  final List<LogLine> lines;

  const EventLogs({
    required this.eventId,
    required this.offset,
    required this.total,
    required this.evicted,
    required this.lines,
  });

  factory EventLogs.fromJson(Map<String, dynamic> json) {
    return EventLogs(
      eventId: json['event_id'] as String,
      offset: (json['offset'] as num).toInt(),
      total: (json['total'] as num).toInt(),
      evicted: json['evicted'] as bool? ?? false,
      lines:
          (json['lines'] as List<dynamic>?)
              ?.map((e) => LogLine.fromJson((e as Map).cast<String, dynamic>()))
              .toList() ??
          const [],
    );
  }

  Map<String, dynamic> toJson() => {
    'event_id': eventId,
    'offset': offset,
    'total': total,
    'evicted': evicted,
    'lines': lines.map((e) => e.toJson()).toList(),
  };
}

class RespondRequest {
  final String eventId;
  final Decision decision;
  final List<String>? selectedOptions;
  final String? textInput;

  const RespondRequest({
    required this.eventId,
    required this.decision,
    this.selectedOptions,
    this.textInput,
  });

  factory RespondRequest.fromJson(Map<String, dynamic> json) {
    return RespondRequest(
      eventId: json['event_id'] as String,
      decision: Decision.fromJson(json['decision'] as String),
      selectedOptions: (json['selected_options'] as List<dynamic>?)
          ?.map((e) => e.toString())
          .toList(),
      textInput: json['text_input'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'event_id': eventId,
    'decision': decision.toJson(),
    if (selectedOptions != null) 'selected_options': selectedOptions,
    if (textInput != null) 'text_input': textInput,
  };
}

class CommandResult {
  final bool ok;
  final CommandError? error;

  const CommandResult({required this.ok, this.error});

  factory CommandResult.fromJson(Map<String, dynamic> json) {
    return CommandResult(
      ok: json['ok'] as bool? ?? false,
      error: json['error'] != null
          ? CommandError.fromJson(json['error'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'ok': ok,
    if (error != null) 'error': error!.toJson(),
  };
}

class ErrorReply {
  final String code;
  final String message;

  const ErrorReply({required this.code, required this.message});

  factory ErrorReply.fromJson(Map<String, dynamic> json) {
    return ErrorReply(
      code: json['code'] as String? ?? 'unknown',
      message: json['message'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {'code': code, 'message': message};
}

/// The envelope carrying every frame.
class Message {
  final String type;
  final String? requestId;
  final dynamic payload;

  const Message({required this.type, this.requestId, this.payload});

  factory Message.push(String type, dynamic payload) {
    return Message(type: type, requestId: null, payload: payload);
  }

  factory Message.withRequestId(
    String type,
    String requestId,
    dynamic payload,
  ) {
    return Message(type: type, requestId: requestId, payload: payload);
  }

  factory Message.request(String requestId, String type, dynamic payload) {
    return Message(type: type, requestId: requestId, payload: payload);
  }

  factory Message.fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String;
    final requestId = json['request_id'] as String?;
    final rawPayload = json['payload'];

    dynamic parsedPayload;
    if (rawPayload is Map) {
      final payloadMap = rawPayload.cast<String, dynamic>();
      switch (type) {
        case 'hello':
          parsedPayload = Hello.fromJson(payloadMap);
          break;
        case 'welcome':
          parsedPayload = Welcome.fromJson(payloadMap);
          break;
        case 'snapshot':
          parsedPayload = Snapshot.fromJson(payloadMap);
          break;
        case 'event':
        case 'event_details':
          parsedPayload = EventPush.fromJson(payloadMap);
          break;
        case 'score_update':
          parsedPayload = ScoreUpdate.fromJson(payloadMap);
          break;
        case 'state_update':
          parsedPayload = StateUpdate.fromJson(payloadMap);
          break;
        case 'raw_event':
          parsedPayload = RawAgentEvent.fromJson(payloadMap);
          break;
        case 'raw_line':
          parsedPayload = RawLine.fromJson(payloadMap);
          break;
        case 'get_event_details':
        case 'ack':
        case 'dismiss':
          parsedPayload = EventRef.fromJson(payloadMap);
          break;
        case 'get_event_logs':
          parsedPayload = GetEventLogs.fromJson(payloadMap);
          break;
        case 'event_logs':
          parsedPayload = EventLogs.fromJson(payloadMap);
          break;
        case 'respond_request':
          parsedPayload = RespondRequest.fromJson(payloadMap);
          break;
        case 'command_result':
          parsedPayload = CommandResult.fromJson(payloadMap);
          break;
        case 'metrics':
          parsedPayload = Map<String, int>.from(
            payloadMap.map((k, v) => MapEntry(k, (v as num).toInt())),
          );
          break;
        case 'error':
          parsedPayload = ErrorReply.fromJson(payloadMap);
          break;
        case 'get_metrics':
          parsedPayload = const {};
          break;
        default:
          parsedPayload = rawPayload;
      }
    } else {
      parsedPayload = rawPayload;
    }

    return Message(type: type, requestId: requestId, payload: parsedPayload);
  }

  Map<String, dynamic> toJson() {
    dynamic serializedPayload;
    if (payload == null) {
      serializedPayload = <String, dynamic>{};
    } else if (payload is Map ||
        payload is List ||
        payload is String ||
        payload is num ||
        payload is bool) {
      serializedPayload = payload;
    } else {
      try {
        serializedPayload = (payload as dynamic).toJson();
      } catch (_) {
        serializedPayload = payload;
      }
    }

    return {
      'type': type,
      if (requestId != null) 'request_id': requestId,
      'payload': serializedPayload,
    };
  }
}

File: core/agentdesk-model/src/message.rs
//! Wire protocol: the envelope and every message type in
//! docs/COMMUNICATION.md. One JSON text frame carries one `Message`.

use std::collections::BTreeMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::event::{AgentId, Event, EventId, RawAgentEvent};
use crate::queue::{EntryState, QueueEntry, Resolution};

/// WebSocket close codes used by the daemon.
pub mod close_code {
    /// Missing/invalid token, or first frame was not `hello`.
    pub const UNAUTHORIZED: u16 = 4001;
    /// Client schema version not supported.
    pub const SCHEMA_MISMATCH: u16 = 4002;
    /// Client could not keep up; outbound channel overflowed.
    pub const SLOW_CLIENT: u16 = 4003;
}

/// Which stages of the laptop pipeline are active. Selected at startup and
/// reported in `welcome`; used by the bench for like-for-like comparison.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PipelineMode {
    RawLines,
    RawEvents,
    Agentdesk,
}

/// How the connection is secured. `InsecureDev` is loopback-only and the
/// phone shows a warning banner when it sees it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TransportMode {
    Tls,
    InsecureDev,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Decision {
    Approve,
    Deny,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CommandError {
    NoSuchEvent,
    NotARequest,
    AlreadyResolved,
    Invalid,
}

/// One raw log line as stored and paged.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LogLine {
    pub offset: u64,
    pub ts: DateTime<Utc>,
    pub text: String,
}

/// Flat counter snapshot; keys are the metric names in ARCHITECTURE.md.
pub type MetricsSnapshot = BTreeMap<String, u64>;

/// Sentinel for `GetEventLogs.offset` meaning "the tail".
pub const LOG_OFFSET_TAIL: i64 = -1;

/// The envelope. `request_id` is present on phone requests and echoed on
/// exactly one reply.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Message {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
    #[serde(flatten)]
    pub body: Body,
}

impl Message {
    pub fn push(body: Body) -> Self {
        Message {
            request_id: None,
            body,
        }
    }
    pub fn with_request_id(request_id: impl Into<String>, body: Body) -> Self {
        Message {
            request_id: Some(request_id.into()),
            body,
        }
    }
}

/// All message types, tagged by `type` with the content under `payload`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload", rename_all = "snake_case")]
pub enum Body {
    // ---- handshake ----
    Hello(Hello),
    Welcome(Welcome),
    Snapshot(Snapshot),

    // ---- laptop → phone push ----
    Event(EventPush),
    ScoreUpdate(ScoreUpdate),
    StateUpdate(StateUpdate),
    RawEvent(RawAgentEvent),
    RawLine(RawLine),

    // ---- phone → laptop requests ----
    GetEventDetails(EventRef),
    GetEventLogs(GetEventLogs),
    Ack(EventRef),
    Dismiss(EventRef),
    RespondRequest(RespondRequest),
    GetMetrics(Empty),

    // ---- laptop → phone replies ----
    EventDetails(EventPush),
    EventLogs(EventLogs),
    CommandResult(CommandResult),
    Metrics(MetricsSnapshot),
    Error(ErrorReply),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Hello {
    pub token: String,
    pub device_id: String,
    pub client_version: String,
    pub schema_version: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Welcome {
    pub daemon_version: String,
    pub schema_version: u32,
    pub pipeline_mode: PipelineMode,
    pub transport: TransportMode,
    pub server_time: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Snapshot {
    pub entries: Vec<QueueEntry>,
    pub events: Vec<Event>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EventPush {
    pub event: Event,
    pub entry: QueueEntry,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ScoreUpdate {
    pub event_id: EventId,
    pub score: u16,
    pub escalation_level: u8,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct StateUpdate {
    pub event_id: EventId,
    pub state: EntryState,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resolution: Option<Resolution>,
    #[serde(default)]
    pub superseded: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RawLine {
    pub agent_id: AgentId,
    #[serde(flatten)]
    pub line: LogLine,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EventRef {
    pub event_id: EventId,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct GetEventLogs {
    pub event_id: EventId,
    /// Per-agent line offset, or `LOG_OFFSET_TAIL`.
    pub offset: i64,
    pub limit: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EventLogs {
    pub event_id: EventId,
    /// Real start offset of `lines`.
    pub offset: u64,
    /// Lines currently retrievable for this event.
    pub total: u64,
    pub evicted: bool,
    pub lines: Vec<LogLine>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RespondRequest {
    pub event_id: EventId,
    pub decision: Decision,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CommandResult {
    pub ok: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<CommandError>,
}

impl CommandResult {
    pub fn ok() -> Self {
        CommandResult {
            ok: true,
            error: None,
        }
    }
    pub fn err(error: CommandError) -> Self {
        CommandResult {
            ok: false,
            error: Some(error),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ErrorReply {
    pub code: String,
    pub message: String,
}

/// Payload for messages that carry nothing. Serialises as `{}`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct Empty {}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::event::*;
    use serde_json::json;
    use uuid::Uuid;

    fn ts() -> DateTime<Utc> {
        "2026-09-17T10:32:05Z".parse().unwrap()
    }

    fn event() -> Event {
        Event {
            schema_version: SCHEMA_VERSION,
            event_id: Uuid::nil(),
            seq: 1,
            agent_seq: 1,
            agent_id: "a".into(),
            agent_name: "A".into(),
            project: "P".into(),
            task_id: Some("t".into()),
            ts: ts(),
            category: Category::Request,
            severity: Severity::Critical,
            kind: "approval_required".into(),
            operation: Operation::Other,
            summary: "Approval required".into(),
            message: "Allow migration?".into(),
            details: Details::new(),
            log_range: LogRange {
                start: 0,
                end: 3,
                pinned: true,
            },
            request: Some(RequestInfo {
                prompt: "Allow migration?".into(),
                options: vec!["approve".into(), "deny".into()],
            }),
        }
    }

    fn entry() -> QueueEntry {
        QueueEntry::new(Uuid::nil(), Category::Request, 1, 90)
    }

    fn line(offset: u64) -> LogLine {
        LogLine {
            offset,
            ts: ts(),
            text: format!("line {offset}"),
        }
    }

    /// One sample of every message type in the catalogue.
    fn all_messages() -> Vec<Message> {
        let id = Uuid::nil();
        vec![
            Message::push(Body::Hello(Hello {
                token: "t".into(),
                device_id: "d".into(),
                client_version: "0.1.0".into(),
                schema_version: SCHEMA_VERSION,
            })),
            Message::push(Body::Welcome(Welcome {
                daemon_version: "0.1.0".into(),
                schema_version: SCHEMA_VERSION,
                pipeline_mode: PipelineMode::Agentdesk,
                transport: TransportMode::Tls,
                server_time: ts(),
            })),
            Message::push(Body::Snapshot(Snapshot {
                entries: vec![entry()],
                events: vec![event()],
            })),
            Message::push(Body::Event(EventPush {
                event: event(),
                entry: entry(),
            })),
            Message::push(Body::ScoreUpdate(ScoreUpdate {
                event_id: id,
                score: 55,
                escalation_level: 1,
            })),
            Message::push(Body::StateUpdate(StateUpdate {
                event_id: id,
                state: EntryState::Seen,
                resolution: Some(Resolution::Unresolved),
                superseded: false,
            })),
            Message::push(Body::RawEvent(RawAgentEvent {
                agent_id: "a".into(),
                agent_seq: 1,
                task_id: None,
                kind: "progress".into(),
                operation: Operation::Build,
                message: "m".into(),
                details: Details::new(),
                log_lines: vec![],
                request: None,
            })),
            Message::push(Body::RawLine(RawLine {
                agent_id: "a".into(),
                line: line(7),
            })),
            Message::with_request_id("r-1", Body::GetEventDetails(EventRef { event_id: id })),
            Message::with_request_id(
                "r-2",
                Body::GetEventLogs(GetEventLogs {
                    event_id: id,
                    offset: LOG_OFFSET_TAIL,
                    limit: 200,
                }),
            ),
            Message::with_request_id("r-3", Body::Ack(EventRef { event_id: id })),
            Message::with_request_id("r-4", Body::Dismiss(EventRef { event_id: id })),
            Message::with_request_id(
                "r-5",
                Body::RespondRequest(RespondRequest {
                    event_id: id,
                    decision: Decision::Approve,
                }),
            ),
            Message::with_request_id("r-6", Body::GetMetrics(Empty {})),
            Message::with_request_id(
                "r-1",
                Body::EventDetails(EventPush {
                    event: event(),
                    entry: entry(),
                }),
            ),
            Message::with_request_id(
                "r-2",
                Body::EventLogs(EventLogs {
                    event_id: id,
                    offset: 5,
                    total: 8,
                    evicted: false,
                    lines: vec![line(5), line(6)],
                }),
            ),
            Message::with_request_id(
                "r-5",
                Body::CommandResult(CommandResult::err(CommandError::AlreadyResolved)),
            ),
            Message::with_request_id(
                "r-6",
                Body::Metrics(MetricsSnapshot::from([("raw_events".to_string(), 42u64)])),
            ),
            Message::with_request_id(
                "r-9",
                Body::Error(ErrorReply {
                    code: "bad_request".into(),
                    message: "cannot parse".into(),
                }),
            ),
        ]
    }

    #[test]
    fn every_message_type_round_trips() {
        let msgs = all_messages();
        assert_eq!(
            msgs.len(),
            19,
            "catalogue size changed; update COMMUNICATION.md"
        );
        for m in msgs {
            let s = serde_json::to_string(&m).unwrap();
            let back: Message = serde_json::from_str(&s).unwrap_or_else(|e| panic!("{e}: {s}"));
            assert_eq!(m, back, "{s}");
        }
    }

    #[test]
    fn envelope_shape_matches_protocol_doc() {
        let m = Message::with_request_id(
            "r-2",
            Body::GetEventLogs(GetEventLogs {
                event_id: Uuid::nil(),
                offset: -1,
                limit: 10,
            }),
        );
        let v = serde_json::to_value(&m).unwrap();
        assert_eq!(v["type"], json!("get_event_logs"));
        assert_eq!(v["request_id"], json!("r-2"));
        assert_eq!(v["payload"]["offset"], json!(-1));
        assert_eq!(v.as_object().unwrap().len(), 3);

        // Push messages omit request_id entirely.
        let p = serde_json::to_value(Message::push(Body::GetMetrics(Empty {}))).unwrap();
        assert!(p.get("request_id").is_none());
        assert_eq!(p["payload"], json!({}));
    }

    #[test]
    fn type_names_are_snake_case_catalogue_names() {
        let names: Vec<String> = all_messages()
            .iter()
            .map(|m| {
                serde_json::to_value(m).unwrap()["type"]
                    .as_str()
                    .unwrap()
                    .to_string()
            })
            .collect();
        assert_eq!(
            names,
            [
                "hello",
                "welcome",
                "snapshot",
                "event",
                "score_update",
                "state_update",
                "raw_event",
                "raw_line",
                "get_event_details",
                "get_event_logs",
                "ack",
                "dismiss",
                "respond_request",
                "get_metrics",
                "event_details",
                "event_logs",
                "command_result",
                "metrics",
                "error",
            ]
        );
    }

    #[test]
    fn unknown_type_is_rejected_and_unknown_fields_ignored() {
        let bad = json!({ "type": "teleport", "payload": {} });
        assert!(serde_json::from_value::<Message>(bad).is_err());

        let ok = json!({ "type": "ack", "request_id": "r", "payload": { "event_id": Uuid::nil(), "extra": 1 }, "trace": "x" });
        let m: Message = serde_json::from_value(ok).unwrap();
        assert_eq!(
            m.body,
            Body::Ack(EventRef {
                event_id: Uuid::nil()
            })
        );
    }

    #[test]
    fn raw_line_flattens_log_line() {
        let v = serde_json::to_value(RawLine {
            agent_id: "a".into(),
            line: line(3),
        })
        .unwrap();
        assert_eq!(v["offset"], json!(3));
        assert_eq!(v["agent_id"], json!("a"));
    }

    #[test]
    fn close_codes() {
        assert_eq!(close_code::UNAUTHORIZED, 4001);
        assert_eq!(close_code::SCHEMA_MISMATCH, 4002);
        assert_eq!(close_code::SLOW_CLIENT, 4003);
    }
}
